pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java

```diff
@@ -48,52 +52,117 @@ public InstanceTagPoolSelector(InstanceTagPoolConfig tagPoolConfig, String table
     if (_tagPoolConfig.isPoolBased()) {
-      // Pool based selection
+      // Pool based selection. All the instances should be associated with a specific pool number.
+      // Instance selection should be done within the same pool.
+      // E.g.: Pool0 -> [ I1, I2, I3 ]
+      //       Pool1 -> [ I4, I5, I6 ]
 
-      // Extract the pool information from the instance configs
-      for (InstanceConfig instanceConfig : candidateInstanceConfigs) {
+      // Each pool number associates with a map that key is the instance name and value is the instance config.
+      Map<Integer, Map<String, InstanceConfig>> poolToInstanceConfigsMap = new HashMap<>();
+      // Each pool number associates with a list of newly added instance configs,
+      // so that new instances can be fetched from this list.
+      Map<Integer, Deque<InstanceConfig>> poolToNewInstanceConfigsMap = new HashMap<>();
+
+      // Extract the pool information from the instance configs.
+      for (Map.Entry<String, InstanceConfig> entry : candidateInstanceConfigsMap.entrySet()) {
+        String instanceName = entry.getKey();
+        InstanceConfig instanceConfig = entry.getValue();
         Map<String, String> poolMap = instanceConfig.getRecord().getMapField(InstanceUtils.POOL_KEY);
         if (poolMap != null && poolMap.containsKey(tag)) {
           int pool = Integer.parseInt(poolMap.get(tag));
-          poolToInstanceConfigsMap.computeIfAbsent(pool, k -> new ArrayList<>()).add(instanceConfig);
+          poolToInstanceConfigsMap.computeIfAbsent(pool, k -> new TreeMap<>()).put(instanceName, instanceConfig);
+          if (newlyAddedInstances.contains(instanceName)) {
+            poolToNewInstanceConfigsMap.computeIfAbsent(pool, k -> new LinkedList<>()).add(instanceConfig);
+          }
+        }
+      }
+
+      for (Map.Entry<Integer, List<String>> entry : existingPoolToInstancesMap.entrySet()) {
+        Integer pool = entry.getKey();
+        List<String> existingInstanceAssignmentInPool = entry.getValue();
+        List<InstanceConfig> candidateInstanceConfigsWithSequence = new ArrayList<>();
+        for (String existingInstance: existingInstanceAssignmentInPool) {
+          InstanceConfig instanceConfig = poolToInstanceConfigsMap.get(pool).get(existingInstance);
+          // Add instances to the candidate list and respect the sequence of the existing instances from the ZK.
+          // The missing/removed instances will be replaced by the newly instances.
+          // If the instance still exists from ZK, then add it to the candidate list.
+          // E.g. if the old instances are: [I1, I2, I3, I4] and the new instance are: [I1, I3, I4, I5, I6],
+          // the removed instance is I2 and the newly added instances are I5 and I6.
+          // The position of I2 would be replaced by I5, the new remaining I6 would be appended to the tail.
+          // Thus, the new order would be [I1, I5, I3, I4, I6].
```

Comment:
`HashBasedRotateInstanceConstraintApplier` will rotate the list which depends on the number of instances, even if we retain the instances order here, after applying the constraints, the order can be different?
Say we need to select 4 instances (2 replicas * 2 instances/replica), and the rotated list is `[I6, I1, I5, I3, I4]`, then we will pick `[I6, I1, I5, I3]`. This won't guarantee minimum segments movement because we need to move all segments from `I4` to `I5` or `I6`. Is this a valid case?

Reply:
Yes this is a valid case, if the pool exists in the existingPoolToInstancesMap, we actually don't need to rotate the list. Otherwise, just do the rotation as the current behavior. Add a class called `RetainedSequenceInstanceConstraintApplier` to extend the existing default applier.

Reply:
`Collections.rotate(instanceConfigs, -(tableNameHash % numInstanceConfigs));`
This does not guarantee the same rotation because `numInstanceConfigs` is different?

Also, even if we maintain the order after we apply the constraints, current `InstanceReplicaGroupPartitionSelector` will pick instances based on # of replica-groups for each pool.
Say we want to change from 2* 2 to 3* 2, the old instance sequence is [I1, I2, I3, I4] and the new sequence is [I1, I5, I3, I4, I6, I7]. The old assignment is:
```
r1 -> {I1, I3}
r2 -> {I2, I4}
```
The new assignment will be:
```
r1 -> {I1, I4}
r2 -> {I5, I6}
r3 -> {I3, I7}
```
But ideally I3 and I4 should not be moved from one replica to another.
We need to improve InstanceReplicaGroupPartitionSelector.selectInstances() to handle such case.

Reply:
The rotation is applied only when it's a new pool. In order to retain the existing instance sequence, we shouldn't rotate the list any more, even if the number of pool has changed.

Reply:
Regarding to your 2nd question, it's a mixed scenario of swapping a host and adding 1 more replica group.
For swapping a host, the current algorithm works well as I5 is going to replace I2 as expected.
For adding 1 more replica group, yes the current algorithm will inevitably shuffle the instances for different replica groups. We can add a TODO here and solve it in another PR.

Reply:
+1 on a follow up PR. I think for the uplift problem we should solve it in/extending InstanceReplicaGroupPartitionSelector.